tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c

```diff
@@ -344,16 +350,19 @@ static void alt_thread_entry(void)
 		"push {r0, r1};\n\t"
 		"mov r1, r7;\n\t"
 		"mov r0, %0;\n\t"
-		"ldmia r0!, {r4-r7};\n\t"
+		"add r0, #16;\n\t"
 		"ldmia r0!, {r4-r7};\n\t"
 		"mov r8, r4;\n\t"
 		"mov r9, r5;\n\t"
 		"mov r10, r6;\n\t"
 		"mov r11, r7;\n\t"
+		"sub r0, #32;\n\t"
+		"ldmia r0!, {r4-r7};\n\t"
 		"mov r7, r1;\n\t"
 		"pop {r0, r1};\n\t"
-		:	: "r" (&ztest_thread_callee_saved_regs_container)
-		: "memory"
+		: /* no output */
+		: "r" (&ztest_thread_callee_saved_regs_container)
+		: "memory", "r4", "r5", "r0", "r1", "r8", "r9", "r10", "r11"
```

Comment:
same deal here

Reply:
If we don't have `r4`, `r5`, etc., in the clobber list, GCC will select `r4` for `%0`, which is what was causing the original problem in this code. We need these values to tell GCC which variables it can safely use for `%0`, which is only `r6` and `r7`. If we remove these, we'll end up in a hard fault.

Reply:
well, this is a hairy situation ... it is playing a dangerous game with the compiler. IMHO, this whole thing needs to be written entirely in asm to guarantee correct behaviour, but that is beyond the scope of this PR ...

Reply:
I agree, the entire function should be written in assembly, but I'd like to at least fix the hard fault here, which this initial rework does.

Reply:
I don't even understand why we're bothering with `ztest_thread_callee_saved_regs_container` if we're just zero'ing the values anyway. There are easier and cleaner solutions for that. But cleaning all of this up should likely be a new PR outside the current bug fix.

Reply:
@gromero Thanks for taking the time to sort some of this out and make a proposal. Happy to test this on HW, and update the PR before we merge.

Reply:
@microbuilder np. Btw, with the "ldmia" form without `!` doesn't work, please revert to your previous code, i.e. change the code pasted above the following:

```
diff --git a/ldmia.c b/ldmia.c
index 81d8fd7..255d90e 100644
--- a/ldmia.c
+++ b/ldmia.c
@@ -66,13 +66,13 @@ int main()
                  */
                 "mov r0, %0;" // since r0 is in the clobber list r0 is not allocated to %0, so r0 != register(%0)
                 "add r0, #16;"
-                "ldmia r0, {r4-r7};"
+                "ldmia r0!, {r4-r7};"
                 "mov r8, r4;"
                 "mov r9, r5;"
                 "mov r10, r6;"
                 "mov r11, r7;"
-                "sub r0, #16;"
-                "ldmia r0, {r4-r7};"
+                "sub r0, #32;"
+                "ldmia r0!, {r4-r7};"
 
                 /*
                  * Restore r0 and r7
```

Reply:
@gromero I noticed that stepping through the code, and looking at the instruction in the armasm PDF for Thumb. Already pushed a fix out, current code should be good for final review.